File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    
    /**
    * Example - 5,7,7,8,8,10 - 
    * mid = 0+5/2 = 2, found 7
    * 7<target, go higher side, 
    * now low = 3, high =5, mid = 3+5/2=4
    * 8 is found
    * recurse - 0, 3 for more 8's
    * mid = 0+3/2 = 1
    * 7<8 -> low = 2, high = 3, mid = 2+3/2 = 2
    * 7<8 -> low = 3, high = 3, mid = 3+3/2 = 3
    * 8 == 8
    *
    * Example - 1,2,3,8,8,*8*,8,8,8,10,12
    * 
    */
     public int[] searchRange(int[] nums, int target) {
            if (nums.length == 0) {
                return new int[]{-1, -1};
            }
            Set<Integer> result = binarySearch(nums, target, 0, nums.length - 1);
            Iterator<Integer> iterator = result.iterator();
            int l = -1;
            int r = -1;

            while (iterator.hasNext()) {
                if (l == -1) {
                    l = iterator.next();
                }
                r = iterator.hasNext() ? iterator.next() : l;
            }
            return new int[]{l, r};
        }
    private Set<Integer> binarySearch(int[] nums, int target, int low, int high) {
            Set<Integer> resultSet = new TreeSet<>();
            while (low <= high) {
                int midIndex = (low + high) / 2;
                if (nums[midIndex] == target) {
                    resultSet.add(midIndex);
                    if (low <= midIndex - 1 && nums[midIndex - 1] == target && (!resultSet.contains(midIndex-1) || !resultSet.contains(low))) {
                        resultSet.addAll(binarySearch(nums, target, low, midIndex - 1));
                    }
                    if (midIndex + 1 <= high && nums[midIndex + 1] == target && (!resultSet.contains(midIndex+1) || !resultSet.contains(high))) {
                        resultSet.addAll(binarySearch(nums, target, midIndex + 1, high));
                    }
                    return resultSet;
                } else if (nums[midIndex] < target) {
                    low = midIndex + 1;
                } else {
                    high = midIndex - 1;
                }
            }
            return resultSet;
        }
}

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.kt
class Solution {
    fun searchRange(nums: IntArray, target: Int): IntArray {
        if (nums.isEmpty()) {
            return intArrayOf(-1, -1)
        }
        return binarySearch(nums, target, 0, nums.size - 1)
    }

    private fun setResult(nums: IntArray, target: Int, midIndex: Int, resultSet: IntArray): Boolean {
        if (midIndex < 0 || midIndex > nums.size - 1) {
            return false
        }
        if(nums[midIndex] != target){
            return false
        }
        var result = false
        if (midIndex < resultSet[0] || resultSet[0] == -1) {
            resultSet[0] = midIndex
            result = true
        }
        if (midIndex > resultSet[1]) {
            resultSet[1] = midIndex
            result = true
        }
        return result
    }
    
    private fun binarySearch(nums: IntArray, target: Int, _low: Int, _high: Int): IntArray {
        //println("Window search $_low and $_high")
        var low = _low
        var high = _high
        if(nums[low] == target && nums[high] == target){
            return intArrayOf(low, high)
        }
        val resultArr = IntArray(2, { -1 })
        while (low <= high) {
            val midIndex = (low + high) / 2
            if (nums[midIndex] == target) {
                val isAddedMin = setResult(nums, target, midIndex - 1, resultArr)
                val isAddedMax = setResult(nums, target, midIndex + 1, resultArr)
                if (!isAddedMin || !isAddedMax) {
                    setResult(nums, target, midIndex, resultArr)
                }
                val lowIndex = midIndex - 2
                val maxIndex = midIndex + 2
                if (low <= lowIndex && nums[lowIndex] == target) {
                    val result = binarySearch(nums, target, low, lowIndex)
                    if (result[0] != -1 && (result[0] < resultArr[0] || resultArr[0] == -1)) {
                        resultArr[0] = result[0]
                    }
                    if (result[1] > resultArr[1]) {
                        resultArr[1] = result[1]
                    }
                }
                if (maxIndex <= high && nums[maxIndex] == target) {
                    val result = binarySearch(nums, target, maxIndex, high)
                    if (result[0] != -1 && (result[0] < resultArr[0] || resultArr[0] == -1)) {
                        resultArr[0] = result[0]
                    }
                    if (result[1] > resultArr[1]) {
                        resultArr[1] = result[1]
                    }
                }
                return resultArr
            } else if (nums[midIndex] < target) {
                low = midIndex + 1
            } else {
                high = midIndex - 1
            }
        }
        return resultArr
    }
}